### PR TITLE
[IMP] mail: allow users to mark messages as unread

### PR DESCRIPTION
**Pull request by sdegueldre**

### CODE REVIEW

File: addons/mail/models/mail_message.py [hunk 1]
```diff
@@ -183,6 +184,11 @@ def _compute_description(self):
                 plaintext_ct = '' if not message.body else tools.html2plaintext(message.body)
                 message.description = plaintext_ct[:30] + '%s' % (' [...]' if len(plaintext_ct) >= 30 else '')
```
Comment (fmdl) [inline addons/mail/models/mail_message.py:186]:
```python
def _compute_is_note(self):
        note_id = self.env['ir.model.data']._xmlid_to_res_id('mail.mt_note')
        for message in self:
            message.is_note = message.subtype_id.id == note_id
```

File: addons/mail/models/mail_channel.py [hunk 2]
```diff
@@ -58,6 +58,12 @@ def _get_default_image(self):
         help="Members of those groups will automatically added as followers. "
              "Note that they will be able to manage their subscription manually "
              "if necessary.")
+    message_unread = fields.Boolean(
+        'Unread Messages', compute='_get_message_unread',
```
Comment (fmdl) [inline addons/mail/models/mail_channel.py:62]: `_get_message_unread` --> `_compute_message_unread`

File: addons/mail/models/mail_message.py [hunk 3]
```diff
@@ -710,6 +710,37 @@ def set_message_done(self):
             }
         )
 
+    def mark_unread(self):
+        """
+        Marks a given message as unread
+        :param last_message_id: the id of the message to be marked as unread.
```
Comment (fmdl) [inline addons/mail/models/mail_message.py:716]: `:param last_message_id` id not present

**Comment by sdegueldre** [comment]
@fmdl this PR is in draft please stop commenting on it until it's actually ready for review, the only reason I'm pushing this is to get the runbot tests and github's diff view. None of this code is ready and half of it is likely to not even be present anymore by the time the PR is marked as ready.

File: addons/mail/models/mail_channel.py [hunk 4]
```diff
@@ -58,6 +58,12 @@ def _get_default_image(self):
...
+        help="If checked, new messages require your attention.")
+    message_unread_counter = fields.Integer(
```
Comment (sdegueldre) [inline addons/mail/models/mail_channel.py:64]: move to channel_partner instead

File: addons/mail/models/mail_channel.py (hunk 2 above)
Comment (seb-odoo) [reply addons/mail/models/mail_channel.py:62]: delete message_unread

File: addons/mail/models/mail_channel.py (hunk 4 above)
Comment (seb-odoo) [reply addons/mail/models/mail_channel.py:64]: and try to make it a simple search count, no query, and not user dependent

File: addons/mail/static/src/components/discuss/tests/discuss_inbox_tests.js [hunk 5]
```diff
@@ -75,7 +75,7 @@ QUnit.test('reply: discard on pressing escape', async function (assert) {
         "should display a single message"
     );
     await afterNextRender(() =>
-        document.querySelector('.o_Message_commandReply').click()
+        document.querySelector('.o_MessageActionList_action.fa-reply').click()
```
Comment (seb-odoo) [inline addons/mail/static/src/components/discuss/tests/discuss_inbox_tests.js:78]: Don't use the icon class a selector, this is prone to error and hard to maintain. Keep a name for the reply button.

File: addons/mail/models/mail_message.py [hunk 6]
```diff
@@ -981,13 +1011,14 @@ def _invalidate_documents(self, model=None, res_id=None):
...
                     'message_ids',
-                    'message_unread',
-                    'message_unread_counter',
                     'message_needaction',
                     'message_needaction_counter',
-                ], ids=[res_id])
+                ]
+                if issubclass(self.pool[model], self.pool['mail.channel']):
+                    fnames += ['message_unread', 'message_unread_counter']
+                self.env[model].invalidate_cache(fnames=fnames, ids=[res_id])
```
Comment (seb-odoo) [inline addons/mail/models/mail_message.py:1021]: this will need to be adapted too then

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 7]
```diff
@@ -1347,20 +1347,25 @@ QUnit.test('basic rendering of message', async function (assert) {
...
-        message.querySelectorAll(`:scope .o_Message_header .o_Message_command`).length,
-        1,
-        "should have a single command in header of message"
+        message.querySelectorAll(`:scope .o_MessageActionList_action`).length,
+        2,
+        "should have two commands in action list of message"
     );
     assert.strictEqual(
-        message.querySelectorAll(`:scope .o_Message_commandStar`).length,
+        message.querySelectorAll(`:scope .o_MessageActionList_action.fa-star-o`).length,
```
Comment (seb-odoo) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:1360]: also keep a name for the action button

File: addons/mail/models/mail_channel.py [hunk 8]
```diff
@@ -113,6 +119,27 @@ def _compute_is_member(self):
         for channel in self:
             channel.is_member = self.env.user.partner_id in channel.channel_partner_ids
 
+    def _get_message_unread(self):
+        partner_id = self.env.user.partner_id.id
+        res = dict.fromkeys(self.ids, 0)
+        if self.ids:
+            # search for unread messages, directly in SQL to improve performances
+            self._cr.execute(""" SELECT msg.res_id FROM mail_message msg
```
Comment (sdegueldre) [inline addons/mail/models/mail_channel.py:127]: use search_count

File: addons/mail/static/src/components/message/message.scss [hunk 9]
```diff
@@ -66,6 +57,7 @@
     margin-top: map-get($spacers, 1);
     margin-bottom: map-get($spacers, 1);
     margin-inline-end: map-get($spacers, 4);
+    flex-grow: 1;
```
Comment (seb-odoo) [inline addons/mail/static/src/components/message/message.scss:60]: could be flew-grow-1 in template

File: addons/mail/static/src/components/message/message.scss [hunk 10]
```diff
@@ -181,12 +150,21 @@
...
+
     &:hover, &.o-clicked {
 
-        .o_Message_commands {
+        .o_MessageActionList {
             opacity: 1;
         }
 
+        .o_Message_contentWrapper {
+            border: 1px solid $border-color;
```
Comment (seb-odoo) [inline addons/mail/static/src/components/message/message.scss:164]: $border-width ?

File: addons/mail/static/src/components/message/message.scss [hunk 11]
```diff
@@ -299,8 +232,12 @@
     }
 }
 
-.o_Message_headerCommands:not(.o-mobile) {
-    opacity: 0;
+.o_Message_contentWrapper {
+    position: relative;
+    border: 1px solid transparent;
+    margin: -1px;
```
Comment (seb-odoo) [inline addons/mail/static/src/components/message/message.scss:238]: I don't like this, I will need to double check in mote details

File: addons/mail/static/src/components/message/message.scss [hunk 12]
```diff
@@ -299,8 +232,12 @@
...
+    margin-left: calc(-#{map-get($spacers, 2)} - 1px);
```
Comment (seb-odoo) [inline addons/mail/static/src/components/message/message.scss:239]: same worry, seems complex

File: addons/mail/static/src/components/message/message.scss [hunk 13]
```diff
@@ -299,8 +232,12 @@
...
+    padding: map-get($spacers, 2);
```
Comment (seb-odoo) [inline addons/mail/static/src/components/message/message.scss:240]: could be p-2 as class

File: addons/mail/static/src/components/message/message.scss [hunk 14]
```diff
@@ -299,8 +232,12 @@
     }
 }
 
-.o_Message_headerCommands:not(.o-mobile) {
-    opacity: 0;
+.o_Message_contentWrapper {
+    position: relative;
```
Comment (seb-odoo) [inline addons/mail/static/src/components/message/message.scss:236]: position-relative

File: addons/mail/static/src/components/message/message.xml [hunk 15]
```diff
@@ -114,75 +101,54 @@
...
-                                    />
-                                </t>
-                            </div>
                         </div>
                     </t>
-                    <div class="o_Message_content" t-ref="content">
-                        <div class="o_Message_prettyBody" t-ref="prettyBody"/><!-- message.prettyBody is inserted here from _update() -->
-                        <t t-if="message.subtype_description and !message.isBodyEqualSubtypeDescription">
-                            <p t-esc="message.subtype_description"/>
+                    <div class="o_Message_contentWrapper">
```
Comment (seb-odoo) [inline addons/mail/static/src/components/message/message.xml:106]: I don't think the wrapper should be around the content, it should probably be the whole message component instead.

File: addons/mail/static/src/components/message/message.xml [hunk 16]
```diff
@@ -114,75 +101,54 @@
...
+                        <MessageActionList
+                            message="message"
```
Comment (seb-odoo) [inline addons/mail/static/src/components/message/message.xml:108]: We had problem in the past giving the reference and not the `localId`, the issue could be somewhat resolved or mitigated now? But I'm not confident in starting to use them. Maybe we can make a pass later.

File: addons/mail/static/src/components/message/message.xml [hunk 17]
```diff
@@ -114,75 +101,54 @@
...
+                            hasReplyIcon="props.hasReplyIcon"
+                            hasMarkAsReadIcon="props.hasReplyIcon"
+                            isReplying="state._wasSelected"
```
Comment (seb-odoo) [inline addons/mail/static/src/components/message/message.xml:111]: I wish all of this was part of a "message view" model.

File: addons/mail/static/src/components/message_action_list/message_action_list.js [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    replyToMessage() {
+        const { message, isReplying } = this.props;
+        if (isReplying) {
+            this.messaging.discuss.clearReplyingToMessage();
+        } else {
+            message.replyTo();
+        }
+    }
```
Comment (seb-odoo) [inline addons/mail/static/src/components/message_action_list/message_action_list.js:32]: move in models all that is possible

File: addons/mail/static/src/components/message_action_list/message_action_list.scss [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+    cursor: pointer;
+
+    &:hover {
+        background-color: mix($border-color, $white);
+    }
+
+    &.fa-star {
+        color: gold;
+    }
+}
```
Comment (seb-odoo) [inline addons/mail/static/src/components/message_action_list/message_action_list.scss:26]: could review css to use some more classes instead, overall need to make a second validation later

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js [hunk 20]
```diff
@@ -546,6 +548,59 @@ function factory(dependencies) {
...
+                    pendingSeenMessageId: undefined,
+                    serverMessageUnreadCounter: channel_unread_counter,
+                });
+            }
+            if (shouldComputeSeenIndicators) {
+                // FIXME force the computing of thread values (cf task-2261221)
+                this.env.models['mail.thread'].computeLastCurrentPartnerMessageSeenByEveryone(channel);
+                // FIXME force the computing of message values (cf task-2261221)
+                this.env.models['mail.message_seen_indicator'].recomputeSeenValues(channel);
+            }
```
Comment (seb-odoo) [inline addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:601]: I think all of this could be finally done with proper compute in models? It was hard before because we had difficulties to express dependencies but it's not longer needed, and we never bothered to clean it.

But I don't want use to duplicate that code  here.

File: addons/mail/static/src/models/thread/thread.js [hunk 21]
```diff
@@ -1374,7 +1374,7 @@ function factory(dependencies) {
             // Include all the messages that are known locally but the server
             // didn't take into account.
             return this.orderedMessages.reduce((total, message) => {
-                if (message.id <= countFromId) {
+                if (message.id <= countFromId || message.isCurrentPartnerAuthor) {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:1377]: What is the reason behind this change? This method is hard to maintain and prone to errors so I don't like diff on it.

Comment (sdegueldre) [reply addons/mail/static/src/models/thread/thread.js:1377]: This is to bring this in line with how message_unread_counter is computed in the backed. Currently, the two don't agree and there is flickering between the two values because of this. I could also change the backend code instead.

File: addons/mail/static/src/components/discuss/tests/discuss_inbox_tests.js (hunk 5 above)
Comment (sdegueldre) [reply addons/mail/static/src/components/discuss/tests/discuss_inbox_tests.js:78]: I find it weird to have classes that only exist to write tests against. I agree that this is not ideal though. Wouldn't it make more sense to use a data-attribute instead? (eg `data-msg-action="reply"`)

File: addons/mail/static/src/components/message/message.xml (hunk 15 above)
Comment (sdegueldre) [reply addons/mail/static/src/components/message/message.xml:106]: That doesn't allow us to place the actions relative to the message text as per the spec.

Comment (seb-odoo) [reply addons/mail/static/src/components/message/message.xml:106]: Why it doesn't allow it? I'm not sure to understand which spec is referred to in that case.
We want the highlight specifically on the content and not on the whole message?

File: addons/mail/static/src/components/discuss/tests/discuss_inbox_tests.js (hunk 5 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_inbox_tests.js:78]: `class` is the standard attribute to describe an element and to allow styling it / finding it, I have no issue with having one even if only for the tests. It also describes the element for many other reasons: makes the code easier to read, allows community to customize it easily, etc.

File: addons/mail/static/src/models/thread/thread.js (hunk 21 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:1377]: I would prefer if the backend one was changed then.
Current user or not should not matter for the count, if a message is marked unread it is should be unread, even if own message.

File: addons/mail/static/src/components/message/message.xml (hunk 15 above)
Comment (sdegueldre) [reply addons/mail/static/src/components/message/message.xml:106]: see first slide of: https://docs.google.com/presentation/d/1WjrTC4YLn_ug8D8OdrGU1HHQe5iAb78U-rHZXrcy8g0/edit#slide=id.gb81872c42a_0_14

Maybe this mockup is not actually what we want? But having played with it a bit I think it makes for a more uniform experience across squashed and non-squashed messages. I'm open to changing to a version where we highlight the entire message instead of just the text and place actions relative to it though.

File: addons/mail/static/src/components/message/message.xml (hunk 17 above)
Comment (sdegueldre) [reply addons/mail/static/src/components/message/message.xml:111]: Not sure what to make of this comment :sweat_smile:

File: addons/mail/models/mail_message.py [hunk 22]
```diff
@@ -710,6 +710,36 @@ def set_message_done(self):
             }
         )
 
+    def mark_unread(self):
+        """
+        Marks the current message as unread.
+        """
+        self.ensure_one()
+        if self.model != 'mail.channel':
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_message.py:718]: Move to channel model. Makes no sense to have a method intended for channels on message model, it should be model agnostic.

File: addons/mail/models/mail_message.py [hunk 23]
```diff
@@ -981,13 +1011,14 @@ def _invalidate_documents(self, model=None, res_id=None):
...
-                self.env[model].invalidate_cache(fnames=[
+                fnames = [
                     'message_ids',
-                    'message_unread',
-                    'message_unread_counter',
                     'message_needaction',
                     'message_needaction_counter',
-                ], ids=[res_id])
+                ]
+                if issubclass(self.pool[model], self.pool['mail.channel']):
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_message.py:1019]: You should probably make a model (mail.thread maybe ?) method instead of hacking message model. Presence of model / res_id parameters indicates this method was located at the wrong place initially.

File: addons/mail/models/mail_thread.py [hunk 24]
```diff
@@ -89,12 +89,6 @@ class MailThread(models.AbstractModel):
         'mail.message', 'res_id', string='Messages',
         domain=lambda self: [('message_type', '!=', 'user_notification')], auto_join=True)
     has_message = fields.Boolean(compute="_compute_has_message", search="_search_has_message", store=False)
-    message_unread = fields.Boolean(
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:92]: Why ? I use it quite frequently. Not sure what you try to gain but I think this kind of fields cost nothing and can be used by people on a daily basis.

File: addons/mail/models/mail_channel.py (hunk 2 above)
Comment (tde-banana-odoo) [reply addons/mail/models/mail_channel.py:62]: I would keep the generic one as it is useful in a daily use.

**Review by tde-banana-odoo — CHANGES_REQUESTED** [review]
Thanks for your PR !

Validation for mail python framework part: some strange removal, code could be improved by making it a bit more modular.

Cheers,

File: addons/mail/models/mail_thread.py (hunk 24 above)
Comment (sdegueldre) [reply addons/mail/models/mail_thread.py:92]: Maybe you'll have to help me understand how it can be used then, as far as I understand the compute does a right-join on channel_partner on channel_id which is a mail.channel, so from my understanding these fields can only ever be False and 0 respectively for non-channel mail.threads.

Although now that I look at it, since we never check that the model is mail.channel, we can probably accidentally select a channel_partner that's completely unrelated to the thread we're interested in, and read an incorrect value for both of these fields because there happens to be a channel_partner for the mail channel with the same id as the thread.

Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py:92]: Seems actually what is called "Message Unread" in interface is using "message_needaction" . My bad, indeed current message_unread makes no sense for standard documents.

I would suggest to actually rename ``message_needaction`` to ``message_unread`` so that naming makes sense. While you are cleaning that part, let us maybe benefits from a renaming :) .

Comment (sdegueldre) [reply addons/mail/models/mail_thread.py:92]: We're sort of getting into a naming spaghetti bowl at this point. In the UI, we have a button called "mark as read" that calls a method named "set_message_done" which removes the "needaction" from the message. On the other hand, channel messages can be "seen" (last_seen_message_id etc) but we want to call the UI button "Mark as unread" because that's what it's called in every other messaging application, so we have this asymmetry between mark read and mark unread which don't represent the same "read" concept at all.

I don't mind cleaning this up to make the naming consistent across the board once the rest of this PR has been validated from a technical standpoint.

Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py:92]: @sdegueldre Well python models are the core of Odoo. I think it is more important to begin with a clear naming for fields, then gradually propagate it. Needaction is an old name coming from FP that meant about nothing. This is linked to unread notifications, so message_unread makes more sense (and 'done' or 'needaction' are wrong naming indeed). Renaming methods and frontend can be done gradually in my opinion. As you prefer anyway, it is your PR :) . Just pointing it out while reading the diff.

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js (hunk 20 above)
Comment (sdegueldre) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:601]: Is this specific to those FIXMEs or are you suggesting that most of what this function does should be implemented through computes?
